### PR TITLE
[SHAPE-UP] Communities (Main branch)

### LINKED ISSUE #4113: Communities Shape-Up (Part I)
We will create a comprehensive solution to integrate communities in Decentraland, allowing users to group around an identity and stay connected with people and events that interest them.

**Design documentation:** https://www.notion.so/decentraland/Communities-Part-I-12a5f41146a580d98a7cfa60cafde8ca

### PR DESCRIPTION
**Pull request by sandrade-dcl**
# Pull Request Description
Fix #4113 

### Prerequisites
- [ ] List any required setup steps
- [ ] Include environment/configuration requirements

### Test Steps
1. First step
2. Second step
3. Expected result after step 2
4. ...

### Additional Testing Notes
- Note any edge cases to verify
- Mention specific areas that need careful testing
- List known limitations or potential issues

## Quality Checklist
- [ ] Changes have been tested locally
- [ ] Documentation has been updated (if required)
- [ ] Performance impact has been considered
- [ ] For SDK features: Test scene is included

## Code Review Reference
Please review our [Code Review Standards](https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md) before submitting.

### CODE REVIEW

**Review by mikhail-dcl — CHANGES_REQUESTED** [review]

File: Explorer/Assets/DCL/Communities/CommunitiesDataProvider/CommunitiesDataProvider.cs [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+            var formData = new List<IMultipartFormSection>
+            {
+                new MultipartFormDataSection("name", name),
+                new MultipartFormDataSection("description", description),
+            };
+
+            List<string> landsAndWorlds = new (lands);
+            landsAndWorlds.AddRange(worlds);
+            StringBuilder placeIdsJsonString = new StringBuilder("[");
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Communities/CommunitiesDataProvider/CommunitiesDataProvider.cs:79]: We shouldn't cache it because the list of lands and worlds varies depending on the community that we're editing.

File: Explorer/Assets/DCL/Communities/CommunitiesDataProvider/CommunitiesDataProvider.cs [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+        {
+            CreateOrUpdateCommunityResponse response;
+
+            var formData = new List<IMultipartFormSection>
+            {
+                new MultipartFormDataSection("name", name),
+                new MultipartFormDataSection("description", description),
+            };
+
+            List<string> landsAndWorlds = new (lands);
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Communities/CommunitiesDataProvider/CommunitiesDataProvider.cs]: Done: https://github.com/decentraland/unity-explorer/pull/4201/commits/92abd8b9da3645742d50122d04b136f358e33520

File: Explorer/Assets/DCL/Communities/CommunityCreation/CommunityCreationEditionController.cs [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+        }
+
+        private void AddCommunityPlace(int index)
+        {
+            if (index >= currentCommunityPlaces.Count)
+                return;
+
+            CommunityPlace selectedPlace = currentCommunityPlaces[index];
+
+            if (addedCommunityPlaces.Exists(place => place.Id == selectedPlace.Id))
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Communities/CommunityCreation/CommunityCreationEditionController.cs]: It was not LINQ but you're right that it was provoking allocations. I have fixed it here: https://github.com/decentraland/unity-explorer/pull/4201/commits/dfc5f1ecd0077e1eb5770e9405ec0b8b5d546135

File: Explorer/Assets/DCL/Communities/CommunityCreation/CommunityCreationEditionController.cs [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+        {
+            viewInstance!.SetCommunityCreationInProgress(true);
+
+            var result = await dataProvider.CreateOrUpdateCommunityAsync(id, name, description, lastSelectedImageData, lands, worlds, ct)
+                                           .SuppressToResultAsync(ReportCategory.COMMUNITIES);
+
+            if (!result.Success)
+            {
+                showErrorCts = showErrorCts.SafeRestart();
+                await viewInstance.WarningNotificationView.AnimatedShowAsync(UPDATE_COMMUNITY_ERROR_MESSAGE, WARNING_MESSAGE_DELAY_MS, showErrorCts.Token);
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Communities/CommunityCreation/CommunityCreationEditionController.cs]: Done: https://github.com/decentraland/unity-explorer/pull/4201/commits/6ac89767fabcd74d08dea83e5f6996e8d52263f0

File: Explorer/Assets/DCL/Communities/CommunitiesFeatureAccess.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
+using Cysharp.Threading.Tasks;
+using DCL.FeatureFlags;
+using DCL.Web3.Identities;
+using System;
+using System.Threading;
+
+namespace DCL.Communities
+{
+    public class CommunitiesFeatureAccess : IDisposable
```
[the remark this answers is not in the document]
Comment (sandrade-dcl) [reply Explorer/Assets/DCL/Communities/CommunitiesFeatureAccess.cs]: Done here: https://github.com/decentraland/unity-explorer/pull/4201/commits/85a264db00d24519809711d9b9db6170adfd4f83

File: Explorer/Assets/DCL/UI/SpriteCache.cs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+        public void AddOrReplaceCachedSprite(string imageUrl, Sprite imageContent)
+        {
+            if(currentSpriteTasks.ContainsKey(imageUrl))
+                currentSpriteTasks[imageUrl].TrySetCanceled();
+
+            if(failedSprites.ContainsKey(imageUrl))
+                failedSprites.Remove(imageUrl);
+
+            if(cachedSprites.ContainsKey(imageUrl))
```
[the remark this answers is not in the document]
Comment (QThund) [reply Explorer/Assets/DCL/UI/SpriteCache.cs]: In the current version of the code, it has been already addressed.

File: Explorer/Assets/DCL/Communities/CommunitiesBrowser/CommunitiesBrowserController.cs [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+        private readonly RectTransform rectTransform;
+        private readonly ICursor cursor;
+        private readonly ICommunitiesDataProvider dataProvider;
+        private readonly IInputBlock inputBlock;
+        private readonly WarningNotificationView warningNotificationView;
+        private readonly IMVCManager mvcManager;
+        private readonly ProfileRepositoryWrapper profileRepositoryWrapper;
+        private readonly ISelfProfile selfProfile;
+        private readonly INftNamesProvider nftNamesProvider;
+        private readonly ObjectProxy<ISpriteCache> spriteCache = new ObjectProxy<ISpriteCache>();
```
[the remark this answers is not in the document]
Comment (QThund) [reply Explorer/Assets/DCL/Communities/CommunitiesBrowser/CommunitiesBrowserController.cs:45]: You are right, I overlooked this. In the case of the browser it is not necessary, but it is in the other cases. I had to use the ObjectProxy becase by the moment the sub-controllers are instantiated, they still don't know which will be the cache, since for example the Community Card can be opened from different places; if it is open from the browser, the browser shares its cache so the card does not need to download anything; if the card is opened from other places it creates its own cache and will destroy its contents when it is closed.

File: Explorer/Assets/DCL/UI/SpriteCache.cs [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+using DCL.WebRequests;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using UnityEngine;
+using Utility;
+
+namespace DCL.UI
+{
+    public class SpriteCache : ISpriteCache
```
[the remark this answers is not in the document]
Comment (QThund) [reply Explorer/Assets/DCL/UI/SpriteCache.cs:13]: I already discussed about this with Juani and Nick. It's true that they do almost the same, but adapting the existing implementations to be used in this scenario was not trivial, as far as I saw, it will require a lot of changes and also modifying the current implementation of the LoadTextureSystem so it is more "generic" and not tight to Avatar textures and so.

File: Explorer/Assets/DCL/Communities/CommunitiesCard/Events/EventListController.cs [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+using PlaceInfo = DCL.PlacesAPIService.PlacesData.PlaceInfo;
+
+namespace DCL.Communities.CommunitiesCard.Events
+{
+    public class EventListController : CommunityFetchingControllerBase<PlaceAndEventDTO, EventListView>
+    {
+        private const int PAGE_SIZE = 20;
+        private const int WARNING_NOTIFICATION_DURATION_MS = 3000;
+
+        private const string LINK_COPIED_MESSAGE = "Link copied to clipboard!";
```
[the remark this answers is not in the document]
Comment (QThund) [reply Explorer/Assets/DCL/Communities/CommunitiesCard/Events/EventListController.cs:31]: In my case it was ignorance, but after asking some people, it seems it's ready to use but it's being barely used throughout our codebase. It would require some learning and some problems may arise while we apply it in real scenarios that involve data coming from the server. I think this requires some research before we start using it in a seriuos way. For incoming shapes, if we really want to add this feature, it should be explicitly mentioned so we all take that into account in our time estimations and in posterior tests.

File: Explorer/Assets/DCL/Communities/CommunitiesCard/Members/CommunityMemberListContextMenuConfiguration.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        [SerializeField] private Sprite addModeratorSprite;
+        [SerializeField] private string addModeratorText = "Add Moderator";
+
+        [SerializeField] private Sprite kickUserSprite;
+        [SerializeField] private string kickUserText = "Kick";
+
+        [SerializeField] private Sprite banUserSprite;
+        [SerializeField] private string banUserText = "Ban";
+
+        public int ContextMenuWidth => contextMenuWidth;
```
[the remark this answers is not in the document]
Comment (lorenzo-ranciaffi) [reply Explorer/Assets/DCL/Communities/CommunitiesCard/Members/CommunityMemberListContextMenuConfiguration.cs]: Good question 🤦

File: Explorer/Assets/DCL/UI/ImageController.cs [hunk 11]
```diff
@@ -13,14 +14,33 @@ public class ImageController
...
+        private readonly IWebRequestController? webRequestController;
+        private CancellationTokenSource cts = new();
         public event Action<Sprite>? SpriteLoaded;
 
+        /// <summary>
+        /// Use this constructor when the image should be stored in a cache.
+        /// </summary>
+        /// <param name="view">The view where the sprite will be presented after loaded.</param>
+        /// <param name="spriteCache">The cache where the sprite will be cached.</param>
+        public ImageController(ImageView view, ObjectProxy<ISpriteCache> spriteCache)
```
[the remark this answers is not in the document]
Comment (QThund) [reply Explorer/Assets/DCL/UI/ImageController.cs:30]: I'm working on removing it...

File: Explorer/Assets/DCL/Communities/CommunitiesBrowser/CommunitiesBrowserController.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (QThund) [reply Explorer/Assets/DCL/Communities/CommunitiesBrowser/CommunitiesBrowserController.cs]: Now ObjectProxy does not exist.

File: Explorer/Assets/DCL/Chat/CommunityChatConversationsToolbarViewItem.cs [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+        /// </summary>
+        /// <param name="thumbnailCache">A way to access thumbnail images asynchronously.</param>
+        /// <param name="imageUrl">The URL to the thumbnail picture.</param>
+        public void SetThumbnailData(ISpriteCache thumbnailCache, string? imageUrl)
+        {
+            customIcon.gameObject.SetActive(false);
+            thumbnailView.SetActive(true);
+
+            if(imageUrl != null)
+                thumbnailView.GetComponent<CommunityThumbnailView>().LoadThumbnailAsync(thumbnailCache, imageUrl, CancellationToken.None).Forget();
```
[the remark this answers is not in the document]
Comment (QThund) [reply Explorer/Assets/DCL/Chat/CommunityChatConversationsToolbarViewItem.cs]: Done.

File: Explorer/Assets/DCL/Chat/MessageBus/MultiplayerChatMessagesBus.cs [hunk 13]
```diff
@@ -21,60 +24,119 @@ public class MultiplayerChatMessagesBus : IChatMessagesBus
...
-            messagePipesHub.IslandPipe().Subscribe<Decentraland.Kernel.Comms.Rfc4.Chat>(Decentraland.Kernel.Comms.Rfc4.Packet.MessageOneofCase.Chat, OnMessageReceived);
-            messagePipesHub.ScenePipe().Subscribe<Decentraland.Kernel.Comms.Rfc4.Chat>(Decentraland.Kernel.Comms.Rfc4.Packet.MessageOneofCase.Chat, OnMessageReceived);
-            messagePipesHub.ChatPipe().Subscribe<Decentraland.Kernel.Comms.Rfc4.Chat>(Decentraland.Kernel.Comms.Rfc4.Packet.MessageOneofCase.Chat, OnPrivateMessageReceived);
+            // Depending on the selected environment, we send the community messages to one user or another
+            string serverEnv = decentralandUrlsSource.Environment == DecentralandEnvironment.Org ? "prd" :
+                               decentralandUrlsSource.Environment == DecentralandEnvironment.Zone ? "dev" :
+                               "local";
+            routingUser = $"message-router-{serverEnv}-0";
+
+            setupExploreSectionsCts = setupExploreSectionsCts.SafeRestart();
```
[the remark this answers is not in the document]
Comment (QThund) [reply Explorer/Assets/DCL/Chat/MessageBus/MultiplayerChatMessagesBus.cs]: Done.

File: Explorer/Assets/DCL/Chat/MessageBus/MultiplayerChatMessagesBus.cs [hunk 14]
```diff
@@ -21,60 +24,119 @@ public class MultiplayerChatMessagesBus : IChatMessagesBus
         private readonly CancellationTokenSource cancellationTokenSource = new ();
         private readonly ObjectProxy<IUserBlockingCache> userBlockingCacheProxy;
         private readonly ChatMessageFactory messageFactory;
+        private readonly string routingUser;
+        private bool isCommunitiesIncluded;
+        private readonly CommunitiesFeatureAccess communitiesFeatureAccess;
+        private CancellationTokenSource setupExploreSectionsCts;
```
[the remark this answers is not in the document]
Comment (QThund) [reply Explorer/Assets/DCL/Chat/MessageBus/MultiplayerChatMessagesBus.cs]: Done.

File: Explorer/Assets/DCL/Communities/CommunityCreation/CommunityCreationEditionController.cs [hunk 15]
```diff
@@ -0,0 +1,419 @@
...
+            viewInstance.SaveCommunityButtonClicked += UpdateCommunity;
+            viewInstance.AddPlaceButtonClicked += AddCommunityPlace;
+            viewInstance.RemovePlaceButtonClicked += RemoveCommunityPlace;
+        }
+
+        protected override void OnBeforeViewShow()
+        {
+            lastSelectedImageData = null;
+            closeTaskCompletionSource = new UniTaskCompletionSource();
+            spriteCache.SetObject(inputData.ThumbnailSpriteCache);
```
[the remark this answers is not in the document]
Comment (QThund) [reply Explorer/Assets/DCL/Communities/CommunityCreation/CommunityCreationEditionController.cs]: ObjectProxy is no more.

File: Explorer/Assets/DCL/UI/ImageController.cs [hunk 16]
```diff
@@ -47,27 +66,43 @@ public async UniTask RequestImageAsync(string uri, bool useKtx, CancellationToke
...
 
-                //TODO potential memory leak, due no CacheCleaner
-                IOwnedTexture2D ownedTexture = await webRequestController.GetTextureAsync(
-                    new CommonArguments(URLAddress.FromString(uri)),
-                    new GetTextureArguments(TextureType.Albedo, useKtx),
-                    GetTextureWebRequest.CreateTexture(TextureWrapMode.Clamp),
-                    ct,
-                    ReportCategory.UI
-                );
+                Sprite? sprite = null;
```
[the remark this answers is not in the document]
Comment (QThund) [reply Explorer/Assets/DCL/UI/ImageController.cs:59]: Done.

**Review by mikhail-dcl — APPROVED** [review]

**Review by DafGreco — APPROVED** [review]
✔️  PR reviewed and approved by QA on both platforms following instructions playing both happy and un-happy path

Regressions for this ticket had been performed in order to verify that the normal flow is working as expected

Some issues were raised but we are going to have a main focus on fixing them on dev ! Congratulations team
